src/ray/stats/metric_exporter.cc

```diff
@@ -110,5 +110,113 @@ void MetricExporter::ExportViewData(
+                                                 boost::asio::io_service &io_service,
+                                                 const std::string address)
+    : client_call_manager_(io_service) {
+  client_.reset(new rpc::MetricsAgentClient(address, port, client_call_manager_));
+};
+
+void OpenCensusProtoExporter::ExportViewData(
+    const std::vector<std::pair<opencensus::stats::ViewDescriptor,
+                                opencensus::stats::ViewData>> &data) {
+  // Start converting opencensus data into their protobuf format.
+  // The format can be found here
+  // https://github.com/census-instrumentation/opencensus-proto/blob/master/src/opencensus/proto/metrics/v1/metrics.proto
+  rpc::ReportOCMetricsRequest request_proto;
+
+  for (const auto &datum : data) {
+    // Unpack the fields we need for in memory data structure.
+    auto view_descriptor = datum.first;
+    auto view_data = datum.second;
+    auto measure_descriptor = view_descriptor.measure_descriptor();
+
+    // Create one metric `Point` in protobuf.
+    auto request_point_proto = request_proto.add_metrics();
+
+    // Write the `MetricDescriptor`.
+    auto metric_descriptor_proto = request_point_proto->mutable_metric_descriptor();
+    metric_descriptor_proto->set_name(measure_descriptor.name());
+    metric_descriptor_proto->set_description(measure_descriptor.description());
+    metric_descriptor_proto->set_unit(measure_descriptor.units());
+    for (const auto &tag_key : view_descriptor.columns()) {
+      metric_descriptor_proto->add_label_keys()->set_key(tag_key.name());
+    };
+
+    // Helpers for writing the actual `TimeSeries`.
+    auto start_time = absl::ToUnixSeconds(view_data.start_time());
+    auto end_time = absl::ToUnixSeconds(view_data.end_time());
+    auto make_new_data_point_proto = [&request_point_proto, start_time, end_time](
+                                         const std::vector<std::string> &tag_values) {
+      auto metric_timeseries_proto = request_point_proto->add_timeseries();
+      metric_timeseries_proto->mutable_start_timestamp()->set_seconds(start_time);
+
+      for (const auto &value : tag_values) {
+        metric_timeseries_proto->add_label_values()->set_value(value);
+      };
+
+      auto point_proto = metric_timeseries_proto->add_points();
+      point_proto->mutable_timestamp()->set_seconds(end_time);
+      return point_proto;
+    };
+
+    // Write the `TimeSeries` for the given aggregated data type.
+    switch (view_data.type()) {
+    case opencensus::stats::ViewData::Type::kDouble:
+      for (const auto &row : view_data.double_data()) {
+        std::vector<std::string> tag_values = row.first;
+
+        auto point_proto = make_new_data_point_proto(tag_values);
+        point_proto->set_double_value(row.second);
+      }
+      break;
+    case opencensus::stats::ViewData::Type::kInt64:
+      for (const auto &row : view_data.int_data()) {
+        auto tag_values = row.first;
```

Comment:
To avoid useless vector copy, we'd better use object reference or remove temp value.